### PR TITLE
Allow asking to avoid mixing hash shorthand styles in single hashes

### PR DESCRIPTION
**Pull request by h-lame**
For those who want to use hash shorthand style, but don't want to mix shorthand and explicit styles in the same hash we provide a new option for `EnforcedShorthandSyntax` called `consistent` that raises an offence if we've mixed styles.

E.g.

    # bad
    {foo: , bar: bar}

    # good
    {foo:, bar:}

    # bad
    {foo: , bar: baz}

    # good
    {foo: foo, bar: baz}

# Requesting help 🙋 

I don't particularly like the implementation.  Prior to this commit `HashSyntax` mixed in `HashShorthandSyntax` but was otherwise unaware of it.  With this change I've had to make `HashSyntax` aware of `HashShorthandSyntax` so it can call `on_hash_for_mixed_shorthand` in `on_hash`.  I don't think there's a way to enforce a no mixing style at the pair level meaning we can't do this in `on_pair` already present in `HashShorthandSyntax` and have to use `on_hash` to look at the whole hash in one go.  Would love some advice on this.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists). (n/a)
* [x] Feature branch is up-to-date with `master` (if not - rebase it). (current rebase against - 0984886d4aacb9b1a6de5c35d5984fac7500d9d0)
* [x] Squashed related commits together. (n/a - single commit)
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/mixin/hash_shorthand_syntax.rb
Comment (jonas054) [inline lib/rubocop/cop/mixin/hash_shorthand_syntax.rb:18]: There should be parentheses here for the method call.

**Review by jonas054 — CHANGES_REQUESTED** [review]
I don't have any problems with the dependency from HashSyntax to the mixin. It all looks good to me. But I'm missing the new supported style in `default.yml`.

File: lib/rubocop/cop/mixin/hash_shorthand_syntax.rb
Comment (h-lame) [reply lib/rubocop/cop/mixin/hash_shorthand_syntax.rb]: Oops.  Thanks for pointing it out!

**Comment by h-lame** [comment]
> I don't have any problems with the dependency from HashSyntax to the mixin. 

Thanks for looking - I'll leave it as-is then!

> But I'm missing the new supported style in `default.yml`.

My bad, I didn't realise I needed to update that too - all done now though!  Rebased again too.

**Review by jonas054 — APPROVED** [review]

**Comment by bbatsov** [comment]
I'm OK with the proposal, but I don't like the name of the config option. I'm think of something like `(keep)_consistent`, but I'm open to other suggestions. Naming is hard. :-)

**Comment by dvandersluis** [comment]
We use `consistent` as a style name elsewhere so it’d be consistent 😂

**Comment by h-lame** [comment]
I admit I'm not sold on `prefer_short` but I thought the `no_mixing` gave some consistency with the `no_mixed_keys` and `ruby19_no_mixed_keys` values for `SupportedStyles`.  Admittedly, `no_mixed_keys` is clear _what_ it's not mixing, I'm not sure what we call this shorthand syntax - it's not keys, and values doesn't feel right to me.  Some alternates:

1. `prefer_short_no_mixing` - what I came up with originally
3. `always_no_mixing` - clear that it's like the value `always` in intent, but has a suffix to suggest a difference (things won't be mixed)
4. `always_keep_consistent` - again, clear that it's like `always` and uses a suffix to suggest a difference (however, it's less clear that there's an implied punctuation between `always` and `keep_consistent` (same problem if we drop `keep` here - `always_consistent` doesn't work either)
5. `always_no_mixed_omissions` - 
6. `for_entire_hashes_only` - going away from all the other options, this feels like it's clearer

I'll keep pondering it 🤔

**Comment by bbatsov** [comment]
Just keep in mind that's not really a supported style, but rather the value of `EnforcedShorthandSyntax`, so the expectation here therefore my suggestion for `consistent`, as everything but "never" implies some usage of the shorthand syntax anyways. I don't like `always_something` as it's not really `always` then.

**Comment by h-lame** [comment]
I don't really have a better suggestion than `consistent`.  What I think it lacks is a hint that it's going to prefer the shorthand syntax and re-write your hashes to use it if it can.  I don't think `consistent` is clear enough.  But, as I said, I don't have a better suggestion, and, as you say, it's kinda implied by the name of the option anyway: `EnforcedShorthandSyntax`.  The docs can explain it too.  I'll rebase and update to use `consistent` as the option value.

File: spec/rubocop/cop/style/hash_syntax_spec.rb
Comment (dvandersluis) [inline spec/rubocop/cop/style/hash_syntax_spec.rb:1302]: `Explicit the hash value` doesn't really make sense.
```suggestion
           ^^^ Do not mix explicit and implicit hash values.  Include the hash value.
```

Comment (h-lame) [reply spec/rubocop/cop/style/hash_syntax_spec.rb:1302]: I agree that it doesn't make much sense, but I was reusing the existing error message for the `never` option.  Should I change that one too - are changes to existing error messages a breaking change?  

We can agree to disagree on the period double space formatting though 😉.  I will remove it though as it does seem more consistent with the rest of the error messages.

Comment (dvandersluis) [reply spec/rubocop/cop/style/hash_syntax_spec.rb]: We could change it as a separate PR if you'd prefer, but if you're willing, change it everywhere please! Changing messages aren't considered breaking.

Comment (h-lame) [reply spec/rubocop/cop/style/hash_syntax_spec.rb]: Added a commit to reword the existing message before adding the new feature.

**Comment by dvandersluis** [comment]
Thanks!

**Comment by koic** [comment]
@dvandersluis From next time onwards, please wait a few days after a minor version release for new features to merge. If there is an urgent bug, it will be a bugfix release. For example, I will wait for @bbatsov to merge new features before merging PR like this. For future reference. Anyway thank you for your activities :-)

**Comment by h-lame** [comment]
@schmijos - I haven't explicitly tested that example, but it should be consistent and not flag for change because the implementation only looks at hashes, and in the "hash" part of that method call (`customer: , store:`) they're already consistent.  If you wanted it to do something else then I think you'd need to add a different implementation around argument lists.

**Pull request merged**
